src/rpc/blockchain.cpp

```diff
@@ -1620,30 +1620,63 @@ UniValue reconsidermostworkchain(const UniValue &params, bool fHelp)
         if (pMostWork->nChainWork < pTip->nChainWork)
             pMostWork = pTip;
     }
+    std::set<CBlockIndex *, CompareBlocksByHeight> setTipsToVerify;
+    setTipsToVerify.insert(pMostWork);
 
-    // If already on the longest chain then return
-    if (pMostWork == chainActive.Tip())
-        throw runtime_error("Nothing to do. Already on the correct chain.");
-
-    // Find where chainActive meets the most work chaintip
-    const CBlockIndex *pFork;
-    pFork = chainActive.FindFork(pMostWork);
-
-    // Rollback to the common forkheight so that both chains will be invalidated.
-    UniValue obj(UniValue::VARR);
-    obj.push_back(pFork->nHeight);
-    if (params.size() > 0)
+    // We need to check if there are duplicate chaintips that have the most work
+    // as could happen during a fork. If there are duplicates then we need to test each tip
+    // to find out which is the correct fork.
     {
-        // Set the rollbackchain override flag if there was one provided.
-        obj.push_back(params[0]);
+        // parse though chaintips again to find if there are duplicates
+        for (CBlockIndex *pTip : setTips)
+        {
+            if (pMostWork->nChainWork == pTip->nChainWork)
+                setTipsToVerify.insert(pTip);
+        }
     }
-    rollbackchain(obj, false);
 
-    // If we got here then rollbackchain() was sucessful and we didn't throw an exception.
-    // Now reconsider the most work chain.
-    UniValue obj_hash(UniValue::VARR);
-    obj_hash.push_back(pMostWork->GetBlockHash().ToString());
-    reconsiderblock(obj_hash, false);
+    for (CBlockIndex *pTipToVerify : setTipsToVerify)
+    {
+        // if no duplicates then return since there is nothing to do. We are already on the correct chain
+        if (pTipToVerify->nChainWork == chainActive.Tip()->nChainWork)
+        {
+            LOGA("Nothing to do. Already on the correct chain.");
+            throw runtime_error("Nothing to do. Already on the correct chain.");
+        }
+
+        // Find where chainActive meets the most work chaintip
+        const CBlockIndex *pFork;
+        pFork = chainActive.FindFork(pTipToVerify);
+
+        // Rollback to the common forkheight so that both chains will be invalidated.
+        UniValue obj(UniValue::VARR);
+        obj.push_back(pFork->nHeight);
+        if (params.size() > 0)
+        {
+            // Set the rollbackchain override flag if there was one provided.
+            obj.push_back(params[0]);
+        }
+        rollbackchain(obj, false);
```

Comment:
RPC call functions should only be called from the RPC interface.  Please make a new function that takes normal arguments, call it here, and have the rollbackchain implementation also call your new function.  (I think there's more than just rollbackchain that needs this to be done).

Reply:
done